numpy/core/src/umath/matmul.c.src

```diff
@@ -494,9 +499,56 @@ NPY_NO_EXPORT void
                  * non-blasable (or non-ccontiguous output)
                  * we could still use BLAS, see gh-12365.
                  */
-                @TYPE@_matmul_inner_noblas(ip1, is1_m, is1_n, 
-                                           ip2, is2_n, is2_p,
-                                           op, os_m, os_p, dm, dn, dp);
+                if(tmp_ip1 == NULL) {
+                    tmp_ip1 = (@typ@ *)malloc(sizeof(@typ@) * dn * dm);
+                }
+                if(tmp_ip2 == NULL) {
+                    tmp_ip2 = (@typ@ *)malloc(sizeof(@typ@) * dn * dp);
+                }
+                if(tmp_op == NULL) {
+                    tmp_op = (@typ@ *)malloc(sizeof(@typ@) * dm * dp);
+                }
+
+                /* Not enough memory */
+                if(
+                    tmp_ip1 == NULL || tmp_ip2 == NULL || tmp_op == NULL
+                ) {
+                    @TYPE@_matmul_inner_noblas(ip1, is1_m, is1_n, 
+                                               ip2, is2_n, is2_p,
+                                               op, os_m, os_p, dm, dn, dp);
+                }
+                else {
+                    for (n = 0; n < dn; n++) {
+                        for (m = 0; m < dm; m++) {
```

Comment:
I didn't check and it is probably true. But you should ensure that `is_m < is_n` (which depends on the memory order).  If that is not true, the loop order needs to be swapped.

Reply:
Can you go a little more into detail what should be done if that is the case?

Reply:
Would probably make sense to cover this with another test case, would it?

Reply:
Well, I guess lapack supports `a.T @ b.T = o.T` mainly, and for that it would make sense to swap things early on (m and n).  It might not matter too much in practice, since it may be that for larger matrices (things don't fit into L1 comfortably?), the actual matmul dominates anyway.

There is one other things which could make sense.  IIRC when `n` is small (mainl?) its not really worthwhile to do the copy at all, very rough timing I think `n < 10` might just be a reasonable thing.